### PR TITLE
feat: Add flag to use operations tags instead of root tags for the sidebar

### PR DESCRIPTION
**Pull request by sebastian-palma**
Implements https://github.com/asyncapi/html-template/issues/42.

`all.js` now defines a function `operationsTags` to get subscription/publish operations tags and use them to build the sidebar.

New functionality available under `sidebarOrganization` param with `byTagsNoRoot` value

### CODE REVIEW

**Comment by derberg** [comment]
In the issue we discussed that best would be to add new parameter or I would add not that best would be to add new value for existing `sidebarOrganization`, so not only have `byTags` but also `byTagsNoRoot`?

when I check your PR with our dummy spec https://github.com/asyncapi/generator/blob/master/test/docs/dummy.yml and I remove root tags from it, it works as expected. We just loose the functionality we had before, the one based on the root tags

Screenshot 2020-06-17 at 14 15 27

**Comment by sebastian-palma** [comment]
Sure @derberg , I forgot that part.

I've added a new function to `all.js` and it made me wonder; is there some way we can test the things we modify in that file? And, do you think we could test the template generation? I mean, adding a new option should change the template, how we could make sure nothing is broken?

I ask from the complete ignorance since I've never worked with Nunjucks. I don't know how you approach these things.

**Comment by derberg** [comment]
@sebastian-palma filters you can easily test as they are normal js functions. Of course, it sucks that we do not have tests yet in this template but we will need them before 1.0 release. When I was working on a library with generic filters I tested most of them, but not all, so you can have a look https://github.com/asyncapi/generator-filters/tree/master/src.

For Nunjucks, yeah, this is beast 😄 we discussed the topic of testing a while ago, for html and markdown template I do not see any other option than snapshot testing. For templates that generate code, I think best is to always have a proper setup with docker compone and integration tests, I do not see any other option.

I'm not focused on Parser, but it will not take more than 2 weeks, thank my next goal is to create a github repository template that will be a template for all the templates (inception 😄 ) so it is easier to start with generator and yeah, have an answer there in the template on how to do testing 😄 it will be a lot of fun

**Comment by sebastian-palma** [comment]
Hey @derberg, it took more time than I thought 😅. I added the new flag for the sidebar organization "byTagsNoRoot" and conditional include the corresponding partials.

I tested with the dummy yaml, and the byTags flag and another yaml file with this new flag and seems to work fine, anyway, I'd appreciate you could try it to see if nothing got broken.

**Comment by derberg** [comment]
@sebastian-palma sorry for the delay man, had some days off, will have a look this week

File: test/spec/asyncapi.yml [hunk 1]
```diff
@@ -39,16 +39,16 @@ channels:
...
+        - name: operation-tag2
           description: Description 2
           externalDocs:
             url: "https://www.asyncapi.com/"
-        - name: oparation-tag3
-        - name: oparation-tag4
+        - name: operation-tag3
+        - name: operation-tag4
           description: Description 4
         - name: message-tag5
```
Comment (derberg) [inline test/spec/asyncapi.yml:53]: can you also remove this, it also should not be here, an old mistake

File: filters/all.js [hunk 2]
```diff
@@ -99,3 +104,15 @@ function containNoTag(channels, tagsToCheck) {
...
+function operationsTags(object) {
+  let tags = new Set();
+  const extractName = (tags, acc) => tags.forEach(({ name }) => acc.add(name));
+  object.channelNames().forEach(channelName => {
+    let channel = object.channel(channelName);
+    if (channel.hasPublish()) extractName(channel._json.publish.tags || [], tags);
+    if (channel.hasSubscribe()) extractName(channel._json.subscribe.tags || [], tags);
+  });
+  return Array.from(tags);
+}
```
Comment (derberg) [inline filters/all.js:117]:
```suggestion
function operationsTags(object) {
  let tags = new Set();
  const extractName = (tags, acc) => tags.forEach((tag) => acc.add(tag.name()));
  object.channelNames().forEach(channelName => {
    let channel = object.channel(channelName);
    if (channel.hasPublish() && channel.publish().hasTags()) extractName(channel.publish().tags(), tags);
    if (channel.hasSubscribe() && channel.subscribe().hasTags()) extractName(channel.subscribe().tags(), tags);
  });
  return Array.from(tags);
}
```

File: filters/all.js [hunk 3]
```diff
@@ -48,7 +49,9 @@ function containTags(object, tagsToCheck) {
     let found = false;
     for (let tagToCheckIndex in tagsToCheck) {
       let tagToCheck = tagsToCheck[tagToCheckIndex]._json;
-      if (tagToCheck.name === tag.name) {
+      let tagName = tag.name;
+      if ((typeof tagToCheck !== 'undefined' && tagToCheck.name === tagName) ||
```
[the remark this answers is not in the document]
Comment (sebastian-palma) [reply filters/all.js:53]: Yep, shorter and better ;)

File: test/spec/asyncapi.yml (hunk 1 above)
Comment (sebastian-palma) [reply test/spec/asyncapi.yml:53]: The `name`s you mean? 👀

Comment (derberg) [reply test/spec/asyncapi.yml:53]: entire tag

Comment (sebastian-palma) [reply test/spec/asyncapi.yml]: There it is. Sorry for the delay.

Comment (derberg) [reply test/spec/asyncapi.yml]: no worries mate, thanks for this additional change but I think I was misunderstood, sorry, by `entire tag` I meant the tag that I commented, so just
```
        - name: message-tag5
          externalDocs:  
            url: "https://www.asyncapi.com/"
```

the rest of tags should stay, sorry again

Comment (sebastian-palma) [reply test/spec/asyncapi.yml]: Cool, no problemo. Just fixed and squashed the commits, thanks @derberg.

**Comment by derberg** [comment]
@all-contributors please add @sebastian-palma for code

**Pull request merged**